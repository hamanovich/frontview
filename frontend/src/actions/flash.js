import { ADD_FLASH_MESSAGE, DELETE_FLASH_MESSAGE, DELETE_FLASH_MESSAGES } from './types';

export const addFlashMessage = payload => ({
  type: ADD_FLASH_MESSAGE,
  payload,
});

export const deleteFlashMessage = id => ({
  type: DELETE_FLASH_MESSAGE,
  id,
});

export const deleteFlashMessages = () => ({
  type: DELETE_FLASH_MESSAGES,
});
