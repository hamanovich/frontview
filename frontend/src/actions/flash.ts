import { Message } from '../propTypes/Message';
import { ADD_FLASH_MESSAGE, DELETE_FLASH_MESSAGE, DELETE_FLASH_MESSAGES } from './types';

interface addFlashMessageAction {
  type: typeof ADD_FLASH_MESSAGE;
  payload: Message;
}

interface deleteFlashMessageAction {
  type: typeof DELETE_FLASH_MESSAGE;
  id: string;
}

interface deleteFlashMessagesAction {
  type: typeof DELETE_FLASH_MESSAGES;
}

export const addFlashMessage = (message: Message) => ({
  type: ADD_FLASH_MESSAGE,
  payload: message,
});

export const deleteFlashMessage = (id: string) => ({
  type: DELETE_FLASH_MESSAGE,
  id,
});

export const deleteFlashMessages = () => ({
  type: DELETE_FLASH_MESSAGES,
});

export type FlashActionTypes =
  | addFlashMessageAction
  | deleteFlashMessageAction
  | deleteFlashMessagesAction;
