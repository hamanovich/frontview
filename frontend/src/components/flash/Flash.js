import React from 'react';
import { func, shape, string } from 'prop-types';
import FontAwesome from 'react-fontawesome';
import MarkdownRenderer from 'react-markdown-renderer';

import classNames from 'classnames';

import compose from 'recompose/compose';
import lifecycle from 'recompose/lifecycle';

import Button from 'react-bootstrap/Button';

import { FlashItem } from './style.ts';

const enhance = compose(
  lifecycle({
    componentDidMount() {
      setTimeout(this.props.close, 4000);
    },
  }),
);

export const Flash = ({ close, message }) => (
  <FlashItem
    variant={classNames({
      danger: message.type === 'error',
      success: message.type === 'success',
      warning: message.type === 'warn',
    })}>
    <MarkdownRenderer markdown={message.text} />
    <Button onClick={close} className="close">
      <FontAwesome name="times" />
    </Button>
  </FlashItem>
);

Flash.propTypes = {
  message: shape({
    type: string,
    text: string,
  }).isRequired,
  close: func.isRequired,
};

export default enhance(Flash);
