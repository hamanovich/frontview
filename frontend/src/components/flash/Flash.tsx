import React, { FC, ReactNode } from 'react';
import FontAwesome from 'react-fontawesome';
import MarkdownRenderer from 'react-markdown-renderer';

import compose from 'recompose/compose';
import lifecycle from 'recompose/lifecycle';

import Button from 'react-bootstrap/Button';

import { FlashProps } from './models';

import { FlashItem } from './style';

const enhance = compose<FlashProps, FlashProps>(
  lifecycle<FlashProps, {}>({
    componentDidMount() {
      setTimeout(this.props.close, 4000);
    },
  }),
);

export const Flash: FC<FlashProps> = ({ close, message }) => {
  const variant =
    message.type === 'error'
      ? 'danger'
      : message.type === 'success'
      ? 'success'
      : 'warning';
  return (
    <FlashItem variant={variant}>
      <MarkdownRenderer markdown={message.text} />
      <Button onClick={close} className="close">
        <FontAwesome name="times" />
      </Button>
    </FlashItem>
  );
};

export default enhance(Flash);
