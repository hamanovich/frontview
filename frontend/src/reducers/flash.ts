import shortid from 'shortid';

import { FlashMessageType } from '../propTypes';

import {
  ADD_FLASH_MESSAGE,
  DELETE_FLASH_MESSAGE,
  DELETE_FLASH_MESSAGES,
} from '../actions/types';
import { FlashActionTypes } from '../actions/flash';

const initialState: FlashMessageType[] = [];

export default (
  state = initialState,
  action: FlashActionTypes,
): FlashMessageType[] => {
  switch (action.type) {
    case ADD_FLASH_MESSAGE: {
      const { type, text } = action.payload;

      return [...state, { id: shortid.generate(), type, text }];
    }

    case DELETE_FLASH_MESSAGE: {
      const index: number = state.findIndex(
        (msg: FlashMessageType) => msg.id === action.id,
      );

      if (index >= 0) {
        return [...state.slice(0, index), ...state.slice(index + 1)];
      }

      return state;
    }

    case DELETE_FLASH_MESSAGES:
      return initialState;

    default:
      return state;
  }
};
