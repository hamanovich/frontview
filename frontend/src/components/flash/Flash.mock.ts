import { action } from '@storybook/addon-actions';
import { FlashProps } from './Flash';

export const mockDataSuccess: FlashProps = {
  close: action('close'),
  message: {
    id: '1',
    type: 'success',
    text: 'Well Done!',
  },
};

export const mockDataWithMarkdown: FlashProps = {
  ...mockDataSuccess,
  message: {
    ...mockDataSuccess.message,
    id: '4',
    text: 'The `Question` with id=__123__ was removed. _Good job_!',
  },
};

export const mockDataWarning: FlashProps = {
  ...mockDataSuccess,
  message: {
    id: '2',
    type: 'warning',
    text: 'Watch out!',
  },
};

export const mockDataDanger: FlashProps = {
  ...mockDataSuccess,
  message: {
    id: '3',
    type: 'error',
    text: 'Be careful!',
  },
};
