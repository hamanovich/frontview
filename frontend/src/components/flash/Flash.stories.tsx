import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, object } from '@storybook/addon-knobs';

import { Flash } from './Flash';
import {
  mockDataSuccess,
  mockDataWarning,
  mockDataDanger,
  mockDataWithMarkdown,
} from './Flash.mock';

storiesOf('Flash|Flash Item', module)
  .addDecorator(withKnobs)
  .add('Success', () => (
    <Flash
      {...mockDataSuccess}
      message={object('message', { ...mockDataSuccess.message })}
    />
  ))
  .add('Warning', () => <Flash {...mockDataWarning} />)
  .add('Danger', () => <Flash {...mockDataDanger} />)
  .add('With Markdown', () => <Flash {...mockDataWithMarkdown} />);
