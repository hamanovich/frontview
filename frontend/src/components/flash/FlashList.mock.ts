import { FlashListProps } from './FlashList';
import { action } from '@storybook/addon-actions';

import { mockDataSuccess, mockDataWarning, mockDataDanger } from './Flash.mock';

export const mockDataDefault: FlashListProps = {
  messages: [
    mockDataSuccess.message,
    mockDataWarning.message,
    mockDataDanger.message,
  ],
  deleteFlashMessage: action('deleteFlashMessage'),
};
