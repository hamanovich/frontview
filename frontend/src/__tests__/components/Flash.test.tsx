import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { Flash, FlashProps } from '../../components/flash/Flash';
import { mockDataDanger } from '../../components/flash/Flash.mock';

describe('<Flash/>', () => {
  const mockData = {
    ...mockDataDanger,
    close: jest.fn(),
  };

  const component: ShallowWrapper<FlashProps> = shallow(
    <Flash {...mockData} />,
  );

  it('renders <Flash /> component', () => {
    expect(component).toMatchSnapshot();
  });

  it('updates via props', () => {
    component.setProps({
      message: { type: 'success', text: 'Success' },
    });

    expect(component.find('MarkdownRenderer').prop('markdown')).toEqual(
      'Success',
    );
  });

  describe('when clicking close button', () => {
    beforeEach(() =>
      component.find('Button').simulate('click', mockData.message.id),
    );

    it('invokes the close callback', () => {
      expect(mockData.close).toHaveBeenCalled();
      expect(mockData.close).toHaveBeenCalledWith(mockData.message.id);
    });
  });
});
