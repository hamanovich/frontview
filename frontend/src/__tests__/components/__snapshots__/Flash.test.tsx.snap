// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flash/> renders <Flash /> component 1`] = `
<Styled(Component)
  variant="danger"
>
  <MarkdownRenderer
    markdown="Error"
  />
  <ForwardRef(Bootstrap(Button))
    className="close"
    onClick={[MockFunction]}
  >
    <FontAwesome
      name="times"
    />
  </ForwardRef(Bootstrap(Button))>
</Styled(Component)>
`;

exports[`<Flash/> updates via props 1`] = `
<Styled(Component)
  variant="success"
>
  <MarkdownRenderer
    markdown="Success"
  />
  <ForwardRef(Bootstrap(Button))
    className="close"
    onClick={[MockFunction]}
  >
    <FontAwesome
      name="times"
    />
  </ForwardRef(Bootstrap(Button))>
</Styled(Component)>
`;
